### DISCUSSION TITLE
About 4th order force constants in ShengBTE format

### DISCUSSION THREAD

**Discussion by ybC007** [ttadano/alamode#93] [General]
Dear Dr. Tadano,

I'm planning to convert the 4th order force constants generated by alm to ShengBTE format. I found that the convert for 3rd force constant is implemented. So I tried to imitate the source code `void Writer::save_fc3_thirdorderpy_format()` to write a new  function for 4th force constants. But when I compiled and run it, it reports the ERROR `ERROR in save_fcs_alamode_oldformat  MESSAGE: This cannot happen.` and I can't find the reason.

I would like to ask that the time when you planning to implement the function. Or would you like to read the code I write for the format convert ?

Thank you very much in advance!
Looking forward to your kind advice.

**Comment by ttadano** [comment]
I have added a new option `FC4_SHENGBTE` in the feature/shengbte_4ph brach. Also see [here](https://github.com/ttadano/alamode/commit/b4c22984faa8d7acda68a66b3fc33ab29b20f5d8).

You can try this version, but please be cautious because I have not tested it.

**Comment by ybC007** [comment]
Cool! I'll have a try.
Thanks a lot!😁

**Comment by ybC007** [comment]
Dear Dr. Tadano,
Hello! I have tried the new option `FC4_SHENGBTE`, but the calculated thermal conductivity  with four phonon interaction seems a little abnormal. Then I read the source code and found there maybe a mistake. But I'm not sure about it. 

In file **writer.cpp** from line 1264 to 1275. 
```
                                        vecs[0][m] = (x_image[0][atom_tmp[0]][m]
                                                      - x_image[0][jat0][m]
                                                      + x_image[cell_now[0]][0][m]
                                                      - x_image[0][0][m]) * Bohr_in_Angstrom;
                                        vecs[1][m] = (x_image[0][atom_tmp[1]][m]
                                                      - x_image[0][kat0][m]
                                                      + x_image[cell_now[1]][0][m]
                                                      - x_image[0][0][m]) * Bohr_in_Angstrom;
                                        vecs[2][m] = (x_image[0][lat][m]
                                                      - x_image[0][lat0][m]
                                                      + x_image[cell_now[2]][0][m]
                                                      - x_image[0][0][m]) * Bohr_in_Angstrom;
```
There may be something with `vecs[2][m] = (x_image[0][lat][m]`? It should probably be modified like this 'vecs[2][m] = (x_image[0][atom_tmp[0]][m]' ?

Thank you very much in advance!
Looking forward to your kind advice.

**Reply by ybC007** [reply]
Sorry for the spelling mistake. It should probably be modified like this `vecs[2][m] = (x_image[0][atom_tmp[2]][m]` ?

**Reply by ttadano** [reply]
Yes, that is obviously a mistake. I've fixed it.
